### PR TITLE
Jwt RSA support

### PR DESCRIPTION
**Pull request by tufanbarisyildirim**
Added RSA support, it uses `config.Secret` as public key when RSA signing method detected. 
Also, updated dependencies.

### CODE REVIEW

**Comment by FZambia** [comment]
@tufanbarisyildirim hello, thanks for pr. In general your pull request looks good (sth wrong with dependencies and some confusion about centrifuge/centrifugo namings).

I had a plan to refactor this part to move away token parse logic from this library. I want this library to know nothing about token parsing logic. Applications that are built on top of this library (including Centrifugo server) will have a way to provide its own function to parse and validate tokens set by clients. Looks like this can be a good time to try doing this.

BTW are you already using RSA tokens so you want Centrifugo server to fit your existing ecosystem? Do you need both RSA and HMAC simultaneously for client tokens or only RSA?

**Comment by tufanbarisyildirim** [comment]
Hi @FZambia , thanks for review, I'm not familiar with it too much, sorry for any confusing and please let me know o I can update the PR. Also yeah it would be good to let apps build their own auth logic. 
Our authentication system supports RSA encryption only and I wanted add RSA support without breaking HMAC compatibility. As far as I tested, it works for both now.

**Comment by FZambia** [comment]
Please give me a time to experiment with moving this out of library during this week, then we'll try to add this into Centrifugo server code. As bonus we will have a chance to name things in a proper way for Centrifugo - let `secret` to be HMAC secret key (can't change name in Centrifugo to prevent breaking backwards compatibility) and introduce sth like `rsa_public_key` for RSA based tokens. This way we will be able to use different secret and public key strings for HMAC and RSA tokens.

**Comment by tufanbarisyildirim** [comment]
That would be great. We will use a custom build until you introduce the new version with RSA. Thank you!

**Comment by tufanbarisyildirim** [comment]
hi again @FZambia what if you just merge and release a new version with that support since it won't break anything?

**Comment by tufanbarisyildirim** [comment]
synced master btw

**Comment by FZambia** [comment]
Currently trying to figure out the best way to refactor this part, I think we at least need separate options for HMAC secret and RSA public key. And also parse public key from string once for RSA ( `ParseRSAPublicKeyFromPEM` looks natural to be called once on start).

I just walked through the code and not sure it's a straightforward task to move away JWT verification part from library - it has some specific error handling and our client libraries handle token expired errors in custom way. Moving this out of library will make contract for custom token parsing pretty complex.

**Comment by tufanbarisyildirim** [comment]
Right, I mean great to plan a big refactoring for the future. Do you want me to separate rsa key and hmac like secret, and public_key? and also move that parsing on startup (and config reload) not  on every call?

**Comment by FZambia** [comment]
> Right, I mean great to plan a big refactoring for the future. Do you want me to separate rsa key and hmac like secret, and public_key? and also move that parsing on startup (and config reload) not on every call?

If you can do this then yeah - will be great. In terms of this library we can still make breaking changes so renaming `Secret` to smth more valuable like `TokenHMACSecretKey` and adding separate `TokenRSAPublicKey` will be right step I suppose. What do you think?

**Comment by tufanbarisyildirim** [comment]
HI @FZambia, updated centrifuge to prefer using `token_hmac_secret_key` if there is and ask for a parsed public (a pkey pointer) instead of a key content. Please review when you have a moment

File: util.go [hunk 1]
```diff
@@ -55,3 +58,26 @@ func putBuffer(buf *bytes.Buffer) {
 	buf.Reset()
 	bufferPool.Put(buf)
 }
+
+func JWTKeyFunc(config Config) func(token *jwt.Token) (interface{}, error) {
```
Comment (FZambia) [inline util.go]: let's make this internal - i.e. call it `jwtKeyFunc`

**Comment by FZambia** [comment]
Thanks!

**Pull request merged**